File: packages/api-models-py/skyportal_py_models/taxonomies.py
"""Request and response models for ``/api/taxonomy``."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from skyportal_py_models.groups import Group


class Taxonomy(BaseModel):
    """A classification taxonomy (upstream ``Taxonomy``).

    ``classifications`` stays untyped: ``classifications.Classification``
    already points at ``Taxonomy``, so typing it would create an import cycle.
    """

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    id: int
    created_at: datetime | None = None
    modified: datetime | None = None
    name: str | None = None
    version: str | None = None
    provenance: str | None = None
    is_latest: bool | None = Field(alias="isLatest", default=None)
    hierarchy: dict[str, Any] | None = None
    groups: list[Group] | None = None
    classifications: list[dict[str, Any]] | None = None


class TaxonomyPost(BaseModel):
    """Payload for creating a taxonomy."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    name: str
    version: str
    hierarchy: dict[str, Any] | None = None
    hierarchy_file: str | None = None
    group_ids: list[int] | None = None
    provenance: str | None = None
    is_latest: bool = Field(alias="isLatest", default=True)


class TaxonomyPut(BaseModel):
    """Payload for updating a taxonomy."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    name: str | None = None
    version: str | None = None
    provenance: str | None = None
    is_latest: bool | None = Field(alias="isLatest", default=None)
    group_ids: list[int] | None = None


class TaxonomyPostResponse(BaseModel):
    """Result of creating a taxonomy."""

    model_config = ConfigDict(extra="forbid")

    taxonomy_id: int

File: packages/api-models-ts/src/Taxonomies.ts
/**
 * Request and response models for `/api/taxonomy`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Groups from "./Groups.ts";
import * as Schemas from "./Schemas.ts";

/**
 * A classification taxonomy (upstream `Taxonomy`).
 *
 * `classifications` stays untyped:
 * {@link skyportal-js/Classifications!Classification} already points at
 * `Taxonomy`, so typing it would create an import cycle.
 *
 * @since 1.0.0
 * @category Models
 */
export const Taxonomy = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        name: Schemas.NullishString,
        version: Schemas.NullishString,
        provenance: Schemas.NullishString,
        isLatest: Schemas.NullishBoolean,
        hierarchy: Schemas.nullish(Schemas.JsonObject),
        groups: Schemas.nullish(v.array(Groups.Group)),
        classifications: Schemas.nullish(v.array(Schemas.JsonObject)),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type Taxonomy = v.InferOutput<typeof Taxonomy>;

/**
 * Payload for creating a taxonomy.
 *
 * Exactly one of `hierarchy` (nested JSON) or `hierarchy_file` (the same
 * structure as a YAML string) must be given, and the hierarchy is validated
 * against the `tdtax` schema. The name/version combination must not already
 * exist. `group_ids` defaults to the public group, and any group the token
 * cannot access is dropped. When `isLatest` is true every other taxonomy with
 * the same name is marked not-latest.
 *
 * @since 1.0.0
 * @category Models
 */
export interface TaxonomyPost {
    readonly name: string;
    readonly version: string;
    readonly hierarchy?: Record<string, unknown> | undefined;
    readonly hierarchy_file?: string | undefined;
    readonly group_ids?: ReadonlyArray<number> | undefined;
    readonly provenance?: string | undefined;
    /** Defaults to true. */
    readonly isLatest?: boolean | undefined;
}

/**
 * Payload for updating a taxonomy.
 *
 * @since 1.0.0
 * @category Models
 */
export interface TaxonomyPut {
    readonly name?: string | undefined;
    readonly version?: string | undefined;
    readonly provenance?: string | undefined;
    readonly isLatest?: boolean | undefined;
    readonly group_ids?: ReadonlyArray<number> | undefined;
}

/**
 * Result of creating a taxonomy.
 *
 * @since 1.0.0
 * @category Models
 */
export const TaxonomyPostResponse = Schemas.model(
    v.strictObject({ taxonomy_id: Schemas.Integer })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type TaxonomyPostResponse = v.InferOutput<typeof TaxonomyPostResponse>;
